### ISSUE TITLE
SMBClient#connect creates shared Connection but does not protect from closing

### ISSUE THREAD

**Issue by matthiasblaesing** [hierynomus/smbj#511]
I saw a strange error in our testing environment. The scenario is this: There is an SMB share, that holds images that are exposed via a REST service on a website. A JSON index is created from the directory listing and thumbnails are created on the fly.

To realize this I originally planned to use one static SMBClient and fetch a connection + authenticate per request (depending on the path different credentials are needed). This lead to strange connection issues. I managed to distill it down to this:

```java
public class TestOpen {
    public static void main(String[] argv) throws Exception {
        String username = "";
        char[] password = "".toCharArray();
        String host = "";
        String share = "";

        try(SMBClient client = new SMBClient(SmbConfig.builder().withDfsEnabled(false).build())) {
            Connection con1 = client.connect(host);
            Connection con2 = client.connect(host);

            Session session1 = con1.authenticate(new AuthenticationContext(username, password, null));
            Session session2 = con2.authenticate(new AuthenticationContext(username, password, null));

            DiskShare share1 = (DiskShare) session1.connectShare(share);
            for(FileIdBothDirectoryInformation fidbdi: share1.list("")) {
                System.out.println(fidbdi.getFileName());
            }
            share1.close();
            session1.close();
            con1.close();

            DiskShare share2 = (DiskShare) session2.connectShare(share);
            for(FileIdBothDirectoryInformation fidbdi: share2.list("")) {
                System.out.println(fidbdi.getFileName());
            }
            share2.close();
            session2.close();
            con2.close();
        }
    }
}
```

This sample should create two directory listings - it produces one log and then fails with a `TransportException`. The log:

```
Nov. 17, 2019 6:03:54 NACHM. com.hierynomus.smbj.connection.Connection connect
INFORMATION: Successfully connected to: XXXXXXXXXXXXXXX
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.hierynomus.security.jce.JceMessageDigest (file:/home/matthias/.m2/repository/com/hierynomus/smbj/0.9.1/smbj-0.9.1.jar) to method sun.security.provider.MD4.getInstance()
WARNING: Please consider reporting this to the maintainers of com.hierynomus.security.jce.JceMessageDigest
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.connection.Connection authenticate
INFORMATION: Successfully authenticated matthias on XXXXXXXXXXXXXXX, session is 2316738742
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.connection.Connection authenticate
INFORMATION: Successfully authenticated matthias on XXXXXXXXXXXXXXX, session is 681239841
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.session.Session connectTree
INFORMATION: Connecting to \\XXXXXXXXXXXXXXX\XXXXXX on session 2316738742
.
..
XXXXXXXX
XXXXX
XXX
XXXXXXXXX
XXXXXXXX
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.session.Session logoff
INFORMATION: Logging off session 2316738742 from host XXXXXXXXXXXXXXX
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.session.Session logoff
INFORMATION: Logging off session 681239841 from host XXXXXXXXXXXXXXX
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.connection.Connection close
INFORMATION: Closed connection to XXXXXXXXXXXXXXX
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.transport.PacketReader run
INFORMATION: Thread[Packet Reader for XXXXXXXXXXXXXXX,5,main] stopped.
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.session.Session connectTree
INFORMATION: Connecting to \\XXXXXXXXXXXXXXX\XXXXXX on session 681239841
Nov. 17, 2019 6:03:55 NACHM. com.hierynomus.smbj.SMBClient close
INFORMATION: Going to close all remaining connections
Exception in thread "main" com.hierynomus.smbj.common.SMBRuntimeException: com.hierynomus.protocol.transport.TransportException: Cannot write SMB2_TREE_CONNECT with message id << 43 >> as transport is disconnected
	at com.hierynomus.smbj.session.Session.connectTree(Session.java:197)
	at com.hierynomus.smbj.session.Session.connectShare(Session.java:144)
	at TestOpen.main(TestOpen.java:32)
Caused by: com.hierynomus.protocol.transport.TransportException: Cannot write SMB2_TREE_CONNECT with message id << 43 >> as transport is disconnected
	at com.hierynomus.smbj.transport.tcp.direct.DirectTcpTransport.write(DirectTcpTransport.java:68)
	at com.hierynomus.smbj.connection.Connection.send(Connection.java:277)
	at com.hierynomus.smbj.session.Session.send(Session.java:285)
	at com.hierynomus.smbj.session.Session.connectTree(Session.java:155)
	... 2 more
```

I think this boils down to:

1. `SMBClient#connect` checks if a connection to a host already exists and if so returns that connection
2. `con1` and `con2` are the same connection
3. When `con1.close()` is called, the connection is shutdown
4. `con2` is now broken as it is the same object

I'm not an expert for SMB, but (1) looks to be the problem here. `Connection` is the physical connection to the server, I would understand, that this should be shared _but_ it is also closeable and I expect every closeable object to be closed if it is not needed anymore. At point (3) above the physical connection should not yet be shut down, as it is shared.

To fix this, I see different options:

1. `SMBClient#connect` could create a new connection per call and leave reuse to the user
2. `SMBClient#connect` could return a proxy object, that tracks usage and only invokes the physical close, if all sharing connections are closed
3. `Connection` could implement usage tracking and let `close` only reduce its `usage` counter, again issuing the physical close after usage count reached zero.

**Comment by hierynomus** [comment]
@matthiasblaesing Yes, if you're reusing the SmbClient from different threads you're getting shared objects that you might not own. 
I'll have a look at the PR

**Comment by matthiasblaesing** [comment]
@hierynomus this is not a threading issue, I just observed it first in a multithreaded environment. The reproducer above is single threaded and reproduces the problem. The call to `session2.connectShare(share)` fails and that fails because the Transport is already disconnected at call time.

**Comment by hierynomus** [comment]
This should be fixed with the closing of #512. Can you confirm @matthiasblaesing

**Comment by matthiasblaesing** [comment]
Yes it works - thank you very much. I had a look at the implementation and it looks safe. I tried to construct race conditions but came up empty. 

One thing I think worth considering is making `Pooled#release` protected. If you have a `Closeable` resource I would always expect users of the resource to end the lifecycle with a call to `close`. With `release` exposed there are two ways to mess with the lease count. Depending on the implementation that derives from `Pooled` `release` might make sense to be exposed, but not always. I'm not sure I agree with the inheritance in this case - this will break if you get more aspects that need to be modelled and I think it is an implementation details that leaks now. But please read this as what it is: Thought after reading your code, not meant to require changes.
